File: contests/coin_change_num_ways.py
from cmath import inf
from typing import Counter
class Solution:
    def coin_change(self):
        coins = self.get_coins()
        target = self.get_target()
        count = [0]*(target + 1)
        
        # If the target is zero, we have got one possible path.
        count[0] = 1
        for curr_target in range(1, target + 1):
            for coin in coins:
                if curr_target - coin >=0:
                    count[curr_target] += count[curr_target - coin]
                    count[curr_target] %= (10**9 + 7)
        return print('\n', count[target])
        
    def get_target(self):
        return int(input())
    def get_coins(self):
        return list(map(int, input().split()))
solution = Solution()
solution.coin_change()

File: contests/coin_change_num_ways.cpp
count[0] = 1; //If the target is zero, we have got one possible path.
for (int x = 1; x <= n; x++) { // loops over all the target from bottom up.
    for (auto c : coins) { // loops over all the coins
        if (x-c >= 0) { // If the current coin is greater than the target, no need to count this path.
            count[x] += count[x-c]; // count the number of ways to reach the target sum
        }
    }
}
cout << count[x] <<endle;
